File: app/src/main/java/repo/dbRepoController/SearchDBRepository.java
package repo.dbRepoController;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.List;

import hlv.cute.todo.App;
import model.Todo;
import repo.dao.TodoDao;

public class SearchDBRepository {

    private final TodoDao dao;
    private final MutableLiveData<List<Todo>> todosLive;

    public SearchDBRepository() {
        dao = App.get().todoDatabase().getTodoDao();
        todosLive = new MutableLiveData<>();
    }

    public void initFetch() {
        new Thread(() -> todosLive.postValue(dao.getAllTodos())).start();
    }

    public void searchTodo(String term) {
        new Thread(() -> todosLive.postValue(dao.searchTodo(term))).start();

    }

    public void searchCategory(String term) {
        new Thread(() -> todosLive.postValue(dao.searchCategory(term))).start();
    }

    public void searchTodoWithCategory(String term) {
        new Thread(() -> todosLive.postValue(dao.searchTodoWithCategory(term))).start();
    }

    public LiveData<List<Todo>> getTodosLive() {
        return todosLive;
    }
}

File: app/src/main/java/repo/dbRepoController/SearchDBRepository.kt
package repo.dbRepoController

import dagger.hilt.android.scopes.ViewModelScoped
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import model.Todo
import repo.dao.SearchDao
import javax.inject.Inject

@ViewModelScoped
class SearchDBRepository @Inject constructor(
    private val dao: SearchDao
) {

    fun getAllTodos(): Flow<List<Todo>?> =
        dao.getAllTodos().distinctUntilChanged()

    suspend fun searchTodo(term: String?): List<Todo>? {
        return dao.searchTodo(term)
    }

    suspend fun searchTodoInSpecificCategory(term: String?, categoryId: Int): List<Todo>? {
        return dao.searchTodoWithCategoryId(term, categoryId)
    }

    suspend fun searchInCategories(term: String?): List<Todo>? {
        return dao.searchCategory(term)
    }

    suspend fun searchInTodosAndCategories(term: String?): List<Todo>? {
        return dao.searchTodoWithCategory(term)
    }

}